app/src/main/java/com/example/spork/home/FetchPlacesData.java

```diff
@@ -0,0 +1,142 @@
+        restaurantList = rs.sortList(restaurantList);
+
+        int max = 10;
+        if (restaurantList.isEmpty()) {
+            max = 0;
+        }
+         else if (restaurantList.size() < 10) {
+             max = restaurantList.size();
+        }
+         for (int i = 0; i < max; i++) {
+             MarkerOptions markerOptions = new MarkerOptions();
+             markerOptions.title(restaurantList.get(i).getName());
+             markerOptions.position(restaurantList.get(i).getLocation());
+             googleMap.addMarker(markerOptions);
+         }
+    }
+
+    @Override
+    protected String doInBackground(Object... objects) {
+
+        try {
+
+            googleMap = (GoogleMap) objects[0];
+            url = (String) objects[1];
+            restaurantList = new ArrayList<>();
+
+            DownloadUrl downloadUrl = new DownloadUrl();
+
+            for (int i = 0; i < 3; i ++) {
+                Log.i(TAG, "current url: " + url);
+
+                String googleNearbyRestaurantsData = downloadUrl.retrieveUrl(url);
+
+                JSONObject jsonObject = new JSONObject(googleNearbyRestaurantsData);
+
+                String status = jsonObject.getString("status");
+
+                if (status.equals("INVALID_REQUEST")) {
+                    for (int k = 1; k < 5; k++) {
+                        Thread.sleep(1000);
+                        Log.i(TAG, "Retrying for invalid request." + k + " times.");
+                        googleNearbyRestaurantsData = downloadUrl.retrieveUrl(url);
+
+                        jsonObject = new JSONObject(googleNearbyRestaurantsData);
+                        status = jsonObject.getString("status");
+                        if (!status.equals("INVALID_REQUEST")) {
+                            break;
+                        }
+                    }
+                }
+
+                JSONArray jsonArray = jsonObject.getJSONArray("results");
+
+                Log.i(TAG, "array size: " + jsonArray.length());
+
+                for (int j = 0; j < jsonArray.length(); j++) {
+                    JSONObject restaurant = jsonArray.getJSONObject(j);
+
+                    JSONObject getLocation = restaurant.getJSONObject("geometry").getJSONObject("location");
+                    String lat = getLocation.getString("lat");
+                    String lng = getLocation.getString("lng");
+                    LatLng latLng = new LatLng(Double.parseDouble(lat), Double.parseDouble(lng));
+
+                    String restaurantName = restaurant.getString("name");
+
+                    boolean isOpen = false;
+
+                    if (restaurant.has("opening_hours")) {
+                        isOpen = Boolean.parseBoolean(restaurant.getJSONObject("opening_hours").getString("open_now"));
+                    }
+
+                    String placeId = restaurant.getString("place_id");
+
+                    if (!restaurant.has("price_level"))
+                        continue;
```

Comment:
Is `price_level` the only possible missing value? And have you decided that if a restaurant doesn't have a price level to ignore it?

Wondering what kinds of restaurants are missing this.

Reply:
For some reason, even when I set the type to restaurant, some places such as movie theatre food courts and catering services get returned. But one distinguishment I noticed between these places is that they often do not have a price_level attribute.